python-packages/smithy-event-stream/smithy_event_stream/aio/interfaces.py

```diff
@@ -0,0 +1,90 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+from typing import Any, Protocol, Self
+
+from smithy_core.deserializers import DeserializeableShape
+from smithy_core.serializers import SerializeableShape
+
+
+class InputEventStream[E: SerializeableShape](Protocol):
+    """Asynchronously sends events to a service."""
+
+    async def send(self, event: E) -> None:
+        """Sends an event to the service.
+
+        :param event: The event to send.
+        """
+        ...
+
+    def close(self) -> None:
+        """Closes the event stream."""
+        ...
+
+    def __enter__(self) -> Self:
+        return self
+
+    def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any):
+        self.close()
+
+
+class OutputEventStream[E: DeserializeableShape](Protocol):
+    """Asynchronously receives events from a service."""
+
+    async def receive(self) -> E | None:
+        """Receive a single event from the service.
+
+        :returns: An event or None. None indicates that no more events will be sent by
+            the service.
+        """
+        ...
+
+    def close(self) -> None:
+        """Closes the event stream."""
+        ...
+
+    async def __anext__(self) -> E:
+        result = await self.receive()
+        if result is None:
+            self.close()
+            raise StopAsyncIteration
+        return result
+
+    def __aiter__(self) -> Self:
+        return self
+
+    def __enter__(self) -> Self:
+        return self
+
+    def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any):
+        self.close()
+
+
+class EventStream[I: InputEventStream[Any] | None, O: OutputEventStream[Any] | None, R](
+    Protocol
+):
+    """A unidirectional or bidirectional event stream."""
+
+    input_stream: I
+    _output_stream: O | None = None
+    _response: R | None = None
+
+    async def await_output(self) -> tuple[R, O]:
+        if self._response is not None:
+            self._response, self._output_stream = await self._await_output()
+
+        return self._response, self._output_stream  # type: ignore
```

Comment:
This chunk I don't know if I fully understand.

We have to establish our input stream outside of the event stream which I'm assuming happens in our `stream` implementation? We don't expect end users to be invoking this?

For the return types, what's in the initial response object vs what we read off the OutputEventStream? Is it just confirmation of the connection and then we're expected to read from the stream?

Reply:
Say you have an operation that returns an event stream. It might have a definition like:

```python
@dataclass(kw_only=True)
class StreamChatsInput:
    room_id: str
    username: str


@dataclass(kw_only=True)
class StreamChatsOutput:
    inactivity_timeout: int
    active_users: list[str]
    

type Chats = ChatMessage
    | ChatReaction
    | ChatResponse
    | ChatLogin
    | ChatLogout
    | ChatKeepalive
    | ChatUnknown


async def stream_chats(self, input: StreamChatsInput) -> EventStream[
    InputEventStream[Chats], OutputEventStream[Chats], StreamChatsOutput]: ...
```

Before this operation returns, it MUST:

* Establish the input stream (`InputEventStream[Chats]`)
* Send the initial request (the contents of `StreamChatsInput`)

The input stream is therefore returned directly by the operation, nested inside the `EventStream` object. Customers may begin sending `Chats` right away.

It MUST NOT:

* Wait for the output stream to be established (`OutputEventStream[Chats]`)
* Wait for the initial response (the contents of `StreamChatsOutput`)

Since the output stream may not have been established and the initial response may not have been received when the operation returns, they are both deffered and must be awaited. The block you highlighted simply caches them so that subequent calls don't have to actually perform any operations on the underlying streams.

As far as what's in the initial response and initial request, it's everything on the top-level input/output that isn't the event stream. In the example above, the initial request dictates what user is connecting and what chat room they're connecting to. Initial messages MUST always be sent first, so the service can rely on that.

Similarly, the initial response MUST be the first message sent in response. This information is likely necessary to understand future messages, or to provide metadata about the stream as established by the service. In the example above it contains information about currently active users and the max amount of time between messages before the service will terminate a connection. This is important information, but an application need not wait on it before sending messages. For example, a chat bot that only establishes a connection to send a single message and then leaves would not necessarily care about that information and thus does not need to wait for the output stream to be established at all.